projects/mci/src/precharge_control.c

```diff
@@ -0,0 +1,145 @@
+ * - Can I keep the precharge state in the motor controller storage, or should I use
+ *       a static variable with a getter method?
+ * - What is the precharge complete message?
+ *    - A: POWER_MAIN_SEQUENCE_PRECHARGE_COMPLETED in EE
+ */
+
+#include "can.h"
+#include "can_ack.h"
+#include "can_msg_defs.h"
+#include "can_transmit.h"
+#include "can_unpack.h"
+#include "gpio.h"
+#include "gpio_it.h"
+#include "status.h"
+
+#include "exported_enums.h"
+
+#include "motor_controller.h"
+
+static const GpioSettings s_control_settings = { .direction = GPIO_DIR_OUT,
+                                                 .state = GPIO_STATE_LOW,
+                                                 .alt_function = GPIO_ALTFN_NONE,
+                                                 .resistor = GPIO_RES_NONE };
+
+static const GpioSettings s_monitor_settings = { .direction = GPIO_DIR_IN,
+                                                 .state = GPIO_STATE_LOW,
+                                                 .alt_function = GPIO_ALTFN_NONE,
+                                                 .resistor = GPIO_RES_NONE };
+static const InterruptSettings s_monitor_it_settings = { .type = INTERRUPT_TYPE_INTERRUPT,
+                                                         .priority = INTERRUPT_PRIORITY_NORMAL };
+
+void prv_monitor_int(const GpioAddress *address, void *context) {
+  MotorControllerStorage *storage = context;
+  GpioState monitor_state = GPIO_STATE_LOW;
+  gpio_get_state(&storage->settings.precharge_monitor, &monitor_state);
+  if (monitor_state == GPIO_STATE_LOW) {
+    storage->precharge_state = MCI_PRECHARGE_DISCHARGED;
+  } else if (monitor_state == GPIO_STATE_HIGH) {
+    storage->precharge_state = MCI_PRECHARGE_CHARGED;
+    CAN_TRANSMIT_POWER_ON_MAIN_SEQUENCE(NULL, EE_POWER_MAIN_SEQUENCE_PRECHARGE_COMPLETED);
+  }
+}
+
+StatusCode prv_precharge(void *context) {
+  MotorControllerStorage *storage = context;
+  // make sure monitor is off
+  GpioState monitor_state = GPIO_STATE_LOW;
+  gpio_get_state(&storage->settings.precharge_monitor, &monitor_state);
+  if (monitor_state != GPIO_STATE_LOW) {
+    return STATUS_CODE_INTERNAL_ERROR;
+  }
+  // set control to on
+  gpio_set_state(&storage->settings.precharge_control, GPIO_STATE_HIGH);
+  gpio_set_state(&storage->settings.precharge_control2, GPIO_STATE_HIGH);
+  // check control to make sure it's on
+  GpioState control_state = GPIO_STATE_LOW;
+  GpioState control2_state = GPIO_STATE_LOW;
+  gpio_get_state(&storage->settings.precharge_control, &control_state);
+  gpio_get_state(&storage->settings.precharge_control2, &control2_state);
+  if (control_state != GPIO_STATE_HIGH || control2_state != GPIO_STATE_HIGH) {
+    return STATUS_CODE_INTERNAL_ERROR;
+  }
+  return STATUS_CODE_OK;
+}
+
+StatusCode prv_discharge(void *context) {
+  // set precharge_control to off
+  MotorControllerStorage *storage = context;
+  gpio_set_state(&storage->settings.precharge_control, GPIO_STATE_LOW);
```

Comment:
okay, I see what you're doing here... but instead of having `precharge_control1` and `precharge_control2`, you can have one `precharge_control`, but then when you initialize this module you can pass in two different `MotorControllerSettings` structs depending on what board you're working with. So you can remove all the double lines here...

Reply:
Micah Black said that hardware accidentally added another precharge control pin and that firmware should be toggling both at the same time